// yaml/browser/dist/schema/common/dist/seq.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.seq = void 0;

var _createNode = require("../../doc/createNode.js");

var _Node = require("../../nodes/Node.js");

var _YAMLSeq = require("../../nodes/YAMLSeq.js");

function createSeq(schema, obj, ctx) {
  var replacer = ctx.replacer;
  var seq = new _YAMLSeq.YAMLSeq(schema);

  if (obj && Symbol.iterator in Object(obj)) {
    var i = 0;
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = obj[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var it = _step.value;

        if (typeof replacer === 'function') {
          var key = obj instanceof Set ? it : String(i++);
          it = replacer.call(obj, key, it);
        }

        seq.items.push((0, _createNode.createNode)(it, undefined, ctx));
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator["return"] != null) {
          _iterator["return"]();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  }

  return seq;
}

var seq = {
  collection: 'seq',
  createNode: createSeq,
  "default": true,
  nodeClass: _YAMLSeq.YAMLSeq,
  tag: 'tag:yaml.org,2002:seq',
  resolve: function resolve(seq, onError) {
    if (!(0, _Node.isSeq)(seq)) onError('Expected a sequence for this tag');
    return seq;
  }
};
exports.seq = seq;
